File: math-engine/cluster_evaluator.py
"""
Sweet Cluster 1000 — Cluster Evaluator
Finds connected clusters of same-symbol cells using Depth-First Search.
This mirrors the TypeScript ClusterEvaluator for RTP validation.
"""

from typing import List, Tuple, Set


def find_clusters(grid: List[List[int]], min_size: int = 5) -> List[dict]:
    """
    Find all connected clusters of matching symbols in the grid.
    
    Args:
        grid: 2D array of symbol IDs. -1 or 7 (scatter) skipped for clustering.
        min_size: Minimum cells in a cluster to count as a win.
    
    Returns:
        List of dicts: { 'symbol_id': int, 'positions': [(row, col), ...] }
    """
    rows = len(grid)
    cols = len(grid[0])
    visited = [[False] * cols for _ in range(rows)]
    clusters = []

    for r in range(rows):
        for c in range(cols):
            if not visited[r][c] and grid[r][c] != -1 and grid[r][c] != 7:
                symbol_id = grid[r][c]
                positions = _dfs(grid, r, c, symbol_id, visited, rows, cols)
                if len(positions) >= min_size:
                    clusters.append({
                        'symbol_id': symbol_id,
                        'positions': positions,
                    })

    return clusters


def _dfs(
    grid: List[List[int]],
    start_r: int,
    start_c: int,
    target: int,
    visited: List[List[bool]],
    rows: int,
    cols: int,
) -> List[Tuple[int, int]]:
    """Depth-first search to find connected cells with the same symbol."""
    stack = [(start_r, start_c)]
    visited[start_r][start_c] = True
    cluster = []

    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # Up, Down, Left, Right

    while stack:
        r, c = stack.pop()
        cluster.append((r, c))

        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and not visited[nr][nc] and grid[nr][nc] == target:
                visited[nr][nc] = True
                stack.append((nr, nc))

    return cluster


def count_scatters(grid: List[List[int]]) -> List[Tuple[int, int]]:
    """Count and locate scatter symbols (ID=7) in the grid."""
    positions = []
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == 7:
                positions.append((r, c))
    return positions

File: src/helpers/ClusterEvaluator.ts
export interface Position {
  row: number;
  col: number;
}

export interface Cluster {
  symbolId: number;
  positions: Position[];
}

export class ClusterEvaluator {
  private grid: number[][]; // 2D array of symbol IDs. -1 means empty.
  private rows: number;
  private cols: number;

  constructor(grid: number[][]) {
    this.grid = grid;
    this.rows = grid.length;
    this.cols = grid[0].length;
  }

  public findClusters(minSize: number = 5): Cluster[] {
    const visited: boolean[][] = Array.from({ length: this.rows }, () =>
      Array(this.cols).fill(false)
    );
    const clusters: Cluster[] = [];

    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.cols; c++) {
        if (!visited[r][c] && this.grid[r][c] !== -1) {
          const symbolId = this.grid[r][c];
          // We ignore scatter (symbolId 8) for cluster logic usually, or handle separately
          if (symbolId === 8) continue; 
          
          // If starting on a Wild, we won't evaluate it by itself as its own cluster.
          // It only evaluates as part of adjacent regular symbols.
          if (symbolId === 7) continue;
          
          const clusterPositions = this.dfs(r, c, symbolId, visited);
          if (clusterPositions.length >= minSize) {
            clusters.push({ symbolId, positions: clusterPositions });
          }
        }
      }
    }

    return clusters;
  }

  private dfs(r: number, c: number, targetSymbol: number, visited: boolean[][]): Position[] {
    const stack: Position[] = [{ row: r, col: c }];
    const clusterPositions: Position[] = [];
    visited[r][c] = true;

    const directions = [
      { dr: -1, dc: 0 }, // Up
      { dr: 1, dc: 0 },  // Down
      { dr: 0, dc: -1 }, // Left
      { dr: 0, dc: 1 }   // Right
    ];

    while (stack.length > 0) {
      const current = stack.pop()!;
      clusterPositions.push(current);

      for (const dir of directions) {
        const nr = current.row + dir.dr;
        const nc = current.col + dir.dc;

        if (this.isValid(nr, nc) && !visited[nr][nc]) {
          const neighborId = this.grid[nr][nc];
          if (neighborId === targetSymbol || neighborId === 7) {
            visited[nr][nc] = true;
            stack.push({ row: nr, col: nc });
          }
        }
      }
    }

    return clusterPositions;
  }

  private isValid(r: number, c: number): boolean {
    return r >= 0 && r < this.rows && c >= 0 && c < this.cols;
  }
}
